src/util/AsioHelpers.h

```diff
@@ -1,65 +1,123 @@
+        auto doCall = [handler = std::move(handler_),
+                       ex = std::current_exception()]() mutable {
+          Value* ptr = nullptr;
+          handler(std::move(ex), std::move(*ptr));
+        };
+        net::post(net::bind_executor(handlerExec, std::move(doCall)));
+      }
+    }
   }
-  AD_CORRECTNESS_CHECK(exceptionPtr);
-  std::rethrow_exception(exceptionPtr);
-}
+};
+// Explicit deduction guides, we need objects and not references as the template
+// parameters.
+template <typename Executor, std::invocable Function, typename Handler>
+CallFunctionAndPassToHandler(Executor&&, Function&&, Handler&&)
+    -> CallFunctionAndPassToHandler<
+        std::decay_t<Executor>, std::decay_t<Function>, std::decay_t<Handler>>;
+}  // namespace detail
 
-// _____________________________________________________________________________
+// Run the `function` on the `executor` (e.g. a strand for synchronization or
+// the executor of a thread pool). As soon as the function has completed, the
+// completion handler which is obtained from the `completionToken` is posted to
+// its associated executor (which might be different from the `executor`).
+template <typename Executor, typename CompletionToken, std::invocable Function>
+auto runFunctionOnExecutor(Executor executor, Function function,
+                           CompletionToken& completionToken) {
+  using Value = std::invoke_result_t<Function>;
+  static constexpr bool isVoid = std::is_void_v<Value>;
 
-/// Helper function that ensures that co_await resumes on the executor
-/// this coroutine was co_spawned on. Overload for void.
-inline net::awaitable<void> resumeOnOriginalExecutor(
-    net::awaitable<void> awaitable) {
-  std::exception_ptr exceptionPtr;
-  try {
-    co_await std::move(awaitable);
-  } catch (...) {
-    exceptionPtr = std::current_exception();
-  }
-  if ((co_await net::this_coro::cancellation_state).cancelled() ==
-      net::cancellation_type::none) {
-    // use_awaitable always resumes the coroutine on the executor the coroutine
-    // was co_spawned on
-    co_await net::post(net::use_awaitable);
-  }
-  if (exceptionPtr) {
-    std::rethrow_exception(exceptionPtr);
-  }
+  // Obtain the call signature for the handler.
+  [[maybe_unused]] auto getSignature = []() {
+    if constexpr (isVoid) {
+      return std::type_identity<void(std::exception_ptr)>{};
+    } else {
+      return std::type_identity<void(std::exception_ptr, Value)>{};
+    }
+  };
+  using Signature = typename decltype(getSignature())::type;
+
+  auto initiatingFunction = [function = std::move(function),
+                             executor](auto&& handler) mutable {
+    auto onExecutor = detail::CallFunctionAndPassToHandler{
+        executor, std::move(function), AD_FWD(handler)};
+    net::post(executor, std::move(onExecutor));
+  };
+  return net::async_initiate<CompletionToken, Signature>(
+      std::move(initiatingFunction), completionToken);
```

Comment:
I'm wondering: What is done here that requires async_initiate not possible with the awaitable interface?

Reply:
I don't quite understand the question.
Is your question why I can't implement this a a qlever combination of `co_spawn` `co_await` and whatnotelse, that requires less code? In that case I am happy to explaing to you next week, why that all doesn't work.

Reply:
Yes, exactly